Framework/API/src/MatrixWorkspace.cpp

```diff
@@ -83,6 +84,31 @@ MatrixWorkspace::~MatrixWorkspace() {
   }
 }
 
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::SpectrumNumbers &&spectrumNumbers,
+    Indexing::DetectorIDs &&detectorIDs) {
+  setIndexTranslator(Indexing::IndexTranslator(std::move(spectrumNumbers),
+                                               std::move(detectorIDs)));
+}
+
+void MatrixWorkspace::setIndexTranslator(
+    Indexing::IndexTranslator &&translator) {
+  if (translator.size() != getNumberHistograms())
+    throw std::runtime_error("MatrixWorkspace::setIndexTranslator: Translator "
+                             "size does not match number of histograms in "
+                             "workspace");
+
+  auto &spectrumNumbers = translator.spectrumNumbers();
+  auto &detectorIDs = translator.detectorIDs();
+  for (size_t i = 0; i < getNumberHistograms(); ++i) {
+    auto &spectrum = getSpectrum(i);
+    spectrum.setSpectrumNo(spectrumNumbers[i]);
+    if (!detectorIDs.empty())
```

Comment:
@OwenArnold So if we want to change only the spectrum numbers and want to avoid partially initialized translator objects, we could do:

```cpp
auto translator = ws.translator; // return by value, cow_ptr inside?
translator.setSpectrumNumbers({1,2,3,4,5});
ws.setTranslator(translator);
```